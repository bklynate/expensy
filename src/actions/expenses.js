/* eslint-disable arrow-body-style */
// import uuid from 'uuid';
import { db } from './../firebase/firebase';

// Action List - Create action generators for these:

// component calls an action generator, which in turn returns an object
// the component then dispatchs an object to the redux store and it changes

// async redux
// component calls action generator
// action generator returns function
// component dispatches function
// function runs and has the ability to cause other actions

// ADD_EXPENSE
export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => dispatch => {
  const {
    description = '',
    note = '',
    amount = 0,
    createdAt = 0
  } = expenseData;
  const expense = { description, note, amount, createdAt };
  return db.ref('expenses').push(expense).then(ref => {
    dispatch(
      addExpense({
        id: ref.key,
        ...expense
      })
    );
  });
};
// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  expense: { id }
});

export const startRemoveExpense = ({ id }) => {
  return (dispatch) => {
    return db.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({ id }))
    })
  }
}
// EDIT_EXPENSE
export const editExpense = ({ id, updates }) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = ({ id, updates }) => {
  return (dispatch) => {
    return db.ref(`expenses/${id}`).update({
      ...updates
    }).then(() => {
      dispatch(editExpense({
        id,
        updates
      }))
    })
  }
}

// SET_EXPENSE
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

export const startSetExpenses = () => {
  return (dispatch) => {
    return db.ref('expenses').once('value')
      .then(data => {
        const expenses = [];
        data.forEach(item => {
          expenses.push({
            id: item.key,
            ...item.val()
          })
        })
        dispatch(setExpenses(expenses));
      })
  }
};
