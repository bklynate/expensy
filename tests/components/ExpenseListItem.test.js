import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from './../../src/components/ExpenseListItem';
import expenses from './../fixtures/expenses';

describe('ExpenseListItem Component', () => {
  it('should render an expense list item', () => {
    const { id, description, amount, createdAt } = expenses[0];

    const wrapper = shallow(
      <ExpenseListItem {...expenses[0]} />
    );

    expect(wrapper).toMatchSnapshot();
  });
});
