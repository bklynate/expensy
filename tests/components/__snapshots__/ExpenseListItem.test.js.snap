// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseListItem Component should render an expense list item 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      Description: 
      November rent payment
    </h3>
  </Link>
  <p>
    Amount: 
    53500
  </p>
  <p>
    Created On: 
    -345600000
  </p>
</div>
`;
